### PR TITLE
Refactor stale-while-revalidate code in HttpCache, add a (first?) test for it

### PR DESCRIPTION
**Pull request by mpdude**
| Q             | A
| ------------- | ---
| Branch?       | 2.8
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

I came up with this while trying to hunt a production bug related to handling of stale cache entries under the condition of a busy backend (also see #22033).

It's just a refactoring to make the code more readable plus a new test.

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php [hunk 1]
```diff
@@ -562,6 +562,42 @@ public function testHitsCachedResponseWithMaxAgeDirective()
...
+
+        // Now, lock the cache
+        $concurrentRequest = Request::create('/', 'GET');
+        $this->store->lock($concurrentRequest);
+
+        /*
+         *  After 10s, the cached response has become stale. Yet, we're still within the "stale_while_revalidate"
+         *  timeout so we may serve the stale response.
+         */
+        sleep(10);
```
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: Sleeping for 20 seconds in total in this test is unacceptable. Is it possible to sleep for less, or don't sleep at all?

Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: With PHPUnitBridge clock mocking, time passes instantily.

Comment (mpdude) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: Doesn't `@group time-sensitive` / clock mocking solve this?

Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: It does indeed. I only checked the method annotations (none). Indeed `time-sensitive` is set on the test class :+1:

**Comment by mpdude** [comment]
@nicolas-grekas should I be concerned because of the failing AppVeyor builds?

**Comment by fabpot** [comment]
Thank you @mpdude.

File: src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php [hunk 2]
```diff
@@ -562,6 +562,42 @@ public function testHitsCachedResponseWithMaxAgeDirective()
...
+         * Another 10s later, stale_while_revalidate is over. Resort to serving the old response, but
+         * do so with a "server unavailable" message.
+         */
+        sleep(10);
+
+        $this->request('GET', '/');
+        $this->assertHttpKernelIsNotCalled();
+        $this->assertEquals(503, $this->response->getStatusCode());
+        $this->assertEquals('Old response', $this->response->getContent());
+    }
```
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: Dunno why, but this test breaks the HttpKernel test suite on windows (see appveyor https://ci.appveyor.com/project/fabpot/symfony/build/1.0.20167#L841), removing the test fixes it.

**Pull request closed without merging**